### DISCUSSION TITLE
How do I look at the contents of a keymap?

### DISCUSSION THREAD

**Discussion by g-gundam** [lem-project/lem#1518] [Q&A] [answered]
Suppose I created a keymap.

```lisp
(defvar *space-b-keymap*
  (make-keymap :name '*space-b-keymap*)
  "buffer menu")
(define-keys *space-b-keymap* 
  ("b" 'lem-core/commands/window:select-buffer)
  ("d" 'lem-core/commands/window:kill-buffer)
  ("n" 'lem-core/commands/window:next-buffer)
  ("p" 'lem-core/commands/window:previous-buffer)
  ("r" 'lem-core/commands/file:revert-buffer))
```

When I try to look at it, I get an opaque representation that isn't very informative.

```
LEM-USER> *space-b-keymap*
#<KEYMAP *SPACE-B-KEYMAP* {100605F933}>
```

What expression could I run on the REPL to get the keys and the functions they're bound to printed?

**Comment by alanz** [comment]
I have been experimenting with [describe-bindings](https://github.com/lem-project/lem/blob/b34d171fcabfb0223053f2bd48189844d96afd18/src/commands/help.lisp#L43), in particular trying to get it to show the `vi-mode` bindings too.  Without success so far.

**Reply by g-gundam** [reply]
Using @jfaz1's answer below, you could input the following into the REPL and click on them to inspect them interactively.
```lisp
lem-vi-mode:*normal-keymap*
lem-vi-mode:*insert-keymap*
lem-vi-mode:*visual-keymap*
lem-vi-mode:*motion-keymap*
lem-vi-mode:*ex-keymap*
lem-vi-mode:*operator-keymap*
```
My `*normal-keymap*` looks like this after I click on the hash table in the `TABLE` slot:
```
#<HASH-TABLE {1001F60A53}>
--------------------
Count: 33
Size: 42
Test: EQ
Rehash size: 1.5
Rehash threshold: 1.0
[clear hashtable]
Contents: 
Space = #<KEYMAP *SPACE-KEYMAP* {1005FBCA13}> [remove entry]
C-n = YANK-POP-NEXT [remove entry]
C-p = YANK-POP [remove entry]
Q = LEM-VI-MODE/COMMANDS:VI-EXECUTE-LAST-RECORDED-MACRO [remove entry]
@ = LEM-VI-MODE/COMMANDS:VI-EXECUTE-MACRO [remove entry]
q = LEM-VI-MODE/COMMANDS:VI-RECORD-MACRO [remove entry]
C-r = LEM-VI-MODE/COMMANDS:VI-REDO [remove entry]
u = LEM-VI-MODE/COMMANDS:VI-UNDO [remove entry]
r = LEM-VI-MODE/COMMANDS:VI-REPLACE-CHAR [remove entry]
P = LEM-VI-MODE/COMMANDS:VI-PASTE-BEFORE [remove entry]
p = LEM-VI-MODE/COMMANDS:VI-PASTE-AFTER [remove entry]
Y = LEM-VI-MODE/COMMANDS:VI-YANK-LINE [remove entry]
y = LEM-VI-MODE/COMMANDS:VI-YANK [remove entry]
J = LEM-VI-MODE/COMMANDS:VI-JOIN-LINE [remove entry]
g = #<HASH-TABLE :TEST EQ :COUNT 4 {1001FFDAC3}> [remove entry]
C = LEM-VI-MODE/COMMANDS:VI-CHANGE-LINE [remove entry]
c = LEM-VI-MODE/COMMANDS:VI-CHANGE [remove entry]
D = LEM-VI-MODE/COMMANDS:VI-DELETE-LINE [remove entry]
d = LEM-VI-MODE/COMMANDS:VI-DELETE [remove entry]
X = LEM-VI-MODE/COMMANDS:VI-DELETE-PREVIOUS-CHAR [remove entry]
x = LEM-VI-MODE/COMMANDS:VI-DELETE-NEXT-CHAR [remove entry]
S = LEM-VI-MODE/COMMANDS:VI-CHANGE-WHOLE-LINE [remove entry]
s = LEM-VI-MODE/COMMANDS:VI-SUBSTITUTE [remove entry]
= = LEM-VI-MODE/COMMANDS:VI-INDENT [remove entry]
. = LEM-VI-MODE/COMMANDS:VI-REPEAT [remove entry]
O = LEM-VI-MODE/COMMANDS:VI-OPEN-ABOVE [remove entry]
o = LEM-VI-MODE/COMMANDS:VI-OPEN-BELOW [remove entry]
A = LEM-VI-MODE/COMMANDS:VI-APPEND-LINE [remove entry]
a = LEM-VI-MODE/COMMANDS:VI-APPEND [remove entry]
I = LEM-VI-MODE/COMMANDS:VI-INSERT-LINE [remove entry]
i = LEM-VI-MODE/COMMANDS:VI-INSERT [remove entry]
Z = #<HASH-TABLE :TEST EQ :COUNT 2 {1001FFD8C3}> [remove entry]
z = #<HASH-TABLE :TEST EQ :COUNT 8 {1001FFD7C3}> [remove entry]
```
What's interesting to me after looking at this for a minute is that the `hjkl` movement keys aren't here.  After looking around a bit more, I found them and many others in `*motion-keymap*`.

**Comment by jfaz1** [comment, marked as answer]
Few ways. Most of the time I just use `describe-bindings` in the mode I'm in. If you want a better way to inspect that specific symbol, you can do `(inspect *space-b-keymap*)` and then view the contents of the table there. Personally, when inspecting I use the `C-c I` keybind hovering over it (or on an empty space and then type the name of the symbol I want to inspect). You can use tab to cycle through the selected item, then RET to open it up.

Also, in your example:

> LEM-USER> *space-b-keymap*
#<KEYMAP *SPACE-B-KEYMAP* {100605F933}>

You can click/hit return over `#<KEYMAP *SPACE-B-KEYMAP* {100605F933}>` to inspect it.

**Reply by g-gundam** [reply]
I didn't know that was clickable.

**Comment by g-gundam** [comment]
## Common Lisp REPL History...  Finally!

By poking around a bit, I finally figured out how to navigate Lem's Common Lisp REPL history.
```lisp
lem/listener-mode::*listener-mode-keymap*
```
...and the keys:
```
#<HASH-TABLE {1001E59633}>
--------------------
Count: 5
Size: 7
Test: EQ
Rehash size: 1.5
Rehash threshold: 1.0
[clear hashtable]
Contents: 
C-c = #<HASH-TABLE :TEST EQ :COUNT 2 {1001FD49C3}> [remove entry]
M-r = LEM/LISTENER-MODE::LISTENER-ISEARCH-HISTORY [remove entry]
M-n = LEM/LISTENER-MODE:LISTENER-NEXT-INPUT [remove entry]
M-p = LEM/LISTENER-MODE:LISTENER-PREVIOUS-INPUT [remove entry]
Return = LEM/LISTENER-MODE:LISTENER-RETURN [remove entry]
```
I was so frustrated by the Common Lisp REPL in Lem  before I found `M-n` and `M-p`.  I found `M-r` by just trying keys out, but for linear history movement, I was trying various combinations of arrow keys and modifier keys.  I was expecting `Up` or `M-Up`, but I never suspected `M-p`.
